Firestore/core/src/firebase/firestore/remote/datastore.mm

```diff
@@ -38,6 +60,102 @@
 }
 
 }  // namespace
+
+const char *const kXGoogAPIClientHeader = "x-goog-api-client";
+const char *const kGoogleCloudResourcePrefix = "google-cloud-resource-prefix";
+
+Datastore::Datastore(util::AsyncQueue *firestore_queue,
+                     const core::DatabaseInfo &database_info)
+    : firestore_queue_{firestore_queue},
+      database_info_{&database_info},
+      grpc_channel_{CreateGrpcChannel()},
+      grpc_stub_{grpc_channel_},
+      dedicated_executor_{CreateExecutor()} {
+  dedicated_executor_->Execute([this] { PollGrpcQueue(); });
+}
+
+void Datastore::Shutdown() {
+  grpc_queue_.Shutdown();
+  // Drain the executor to make sure it extracted all the operations from gRPC
+  // completion queue.
+  dedicated_executor_->ExecuteBlocking([] {});
+}
+
+void Datastore::PollGrpcQueue() {
+  HARD_ASSERT(dedicated_executor_->IsCurrentExecutor(),
+              "PollGrpcQueue should only be called on the "
+              "dedicated Datastore executor");
+
+  void *tag = nullptr;
+  bool ok = false;
+  while (grpc_queue_.Next(&tag, &ok)) {
+    auto operation = static_cast<GrpcOperation *>(tag);
+    HARD_ASSERT(tag, "GRPC queue returned a null tag");
+    operation->Complete(ok);
+  }
+}
+
+std::shared_ptr<grpc::Channel> Datastore::CreateGrpcChannel() const {
+  return grpc::CreateChannel(
+      database_info_->host(),
+      grpc::SslCredentials(grpc::SslCredentialsOptions()));
+}
+
+void Datastore::EnsureValidGrpcStub() {
+  // TODO(varconst): find out in which cases a gRPC channel might shut down.
+  // This might be overkill.
+  if (!grpc_channel_ ||
+      grpc_channel_->GetState(false) == GRPC_CHANNEL_SHUTDOWN) {
+    grpc_channel_ = CreateGrpcChannel();
+    grpc_stub_ = grpc::GenericStub{grpc_channel_};
+  }
+}
+
+std::unique_ptr<GrpcStream> Datastore::CreateGrpcStream(
+    absl::string_view token,
+    absl::string_view path,
+    GrpcStreamObserver *observer) {
+  EnsureValidGrpcStub();
+
+  auto context = CreateGrpcContext(token);
+  auto reader_writer = CreateGrpcReaderWriter(context.get(), path);
+  return absl::make_unique<GrpcStream>(
+      std::move(context), std::move(reader_writer), observer, firestore_queue_);
+}
+
+std::unique_ptr<grpc::ClientContext> Datastore::CreateGrpcContext(
+    absl::string_view token) const {
+  auto context = absl::make_unique<grpc::ClientContext>();
+  if (token.data()) {
+    context->set_credentials(grpc::AccessTokenCredentials(token.data()));
+  }
+
+  // TODO(dimond): This should ideally also include the grpc version, however,
+  // gRPC defines the version as a macro, so it would be hardcoded based on
+  // version we have at compile time of the Firestore library, rather than the
+  // version available at runtime/at compile time by the user of the library.
+  //
+  // TODO(varconst): once C++ SDK is ready, this should be "gl-cpp" or similar.
```

Comment:
Once the C++ SDK is ready this likely should be parameterized somehow so that we can see if the client is Unity, C++, or Objective-C.